From b9de3bd85fae8070483320759e55a75201a922c2 Mon Sep 17 00:00:00 2001
From: Aaron He <aaron@bluestacks.com>
Date: Sat, 23 Sep 2023 15:06:30 +0800
Subject: [PATCH 13/21] [A13] stagefright: Add support for loading a custom
 OMXPlugin

 * To facilitate moving the stagefright-plugins glue out of the
   framework, support is added to OMXMaster to load multiple
   external plugins besides internal/vendor versions.
 * This is currently limited to one plugin, defined by the
   "mm.sf.omx-plugin" system property. The code will allow any
   number of libraries to be loaded, though.
 * Should also be useful for nonstandard vendor implementations too.
---
 media/libstagefright/omx/OMXStore.cpp                      | 7 +++++++
 media/libstagefright/omx/SoftOMXPlugin.cpp                 | 6 +++++-
 .../omx/include/media/stagefright/omx/OMXStore.h           | 1 +
 3 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/media/libstagefright/omx/OMXStore.cpp b/media/libstagefright/omx/OMXStore.cpp
index 930fbb7fb0..243a7df30a 100644
--- a/media/libstagefright/omx/OMXStore.cpp
+++ b/media/libstagefright/omx/OMXStore.cpp
@@ -55,6 +55,7 @@ OMXStore::OMXStore() {
 
     addVendorPlugin();
     addPlatformPlugin();
+    addUserPlugin();
 }
 
 OMXStore::~OMXStore() {
@@ -69,6 +70,12 @@ void OMXStore::addPlatformPlugin() {
     addPlugin("libstagefright_softomx_plugin.so");
 }
 
+void OMXStore::addUserPlugin() {
+    std::string plugin = android::base::GetProperty("media.sf.omx-plugin", "");
+    ALOGD("user omx plugin: %s\n", plugin.c_str());
+    addPlugin(plugin.c_str());
+}
+
 void OMXStore::addPlugin(const char *libname) {
     if (::android::base::GetIntProperty("vendor.media.omx", int64_t(1)) == 0) {
         return;
diff --git a/media/libstagefright/omx/SoftOMXPlugin.cpp b/media/libstagefright/omx/SoftOMXPlugin.cpp
index 8c186c90f8..e93c8528be 100644
--- a/media/libstagefright/omx/SoftOMXPlugin.cpp
+++ b/media/libstagefright/omx/SoftOMXPlugin.cpp
@@ -86,6 +86,7 @@ OMX_ERRORTYPE SoftOMXPlugin::makeComponentInstance(
         OMX_COMPONENTTYPE **component) {
     ALOGV("makeComponentInstance '%s'", name);
 
+    dlerror(); // clear any existing error
     for (size_t i = 0; i < kNumComponents; ++i) {
         if (strcmp(name, kComponents[i].mName)) {
             continue;
@@ -117,6 +118,8 @@ OMX_ERRORTYPE SoftOMXPlugin::makeComponentInstance(
             return OMX_ErrorComponentNotFound;
         }
 
+        ALOGV("load component %s for %s", libName.c_str(), name);
+
         typedef SoftOMXComponent *(*CreateSoftOMXComponentFunc)(
                 const char *, const OMX_CALLBACKTYPE *,
                 OMX_PTR, OMX_COMPONENTTYPE **);
@@ -127,7 +130,8 @@ OMX_ERRORTYPE SoftOMXPlugin::makeComponentInstance(
                     "_Z22createSoftOMXComponentPKcPK16OMX_CALLBACKTYPE"
                     "PvPP17OMX_COMPONENTTYPE");
 
-        if (createSoftOMXComponent == NULL) {
+        if (const char *error = dlerror()) {
+            ALOGE("unable to dlsym %s: %s", libName.c_str(), error);
             dlclose(libHandle);
             libHandle = NULL;
 
diff --git a/media/libstagefright/omx/include/media/stagefright/omx/OMXStore.h b/media/libstagefright/omx/include/media/stagefright/omx/OMXStore.h
index 5d6c3ed29d..aaaefdaaad 100644
--- a/media/libstagefright/omx/include/media/stagefright/omx/OMXStore.h
+++ b/media/libstagefright/omx/include/media/stagefright/omx/OMXStore.h
@@ -62,6 +62,7 @@ private:
 
     void addVendorPlugin();
     void addPlatformPlugin();
+    void addUserPlugin();
     void addPlugin(const char *libname);
     void addPlugin(OMXPluginBase *plugin);
     void clearPlugins();
